### PR TITLE
Global app switcher mf 430

### PR DESCRIPTION
**Pull request by joachim-junior**
This ticket was about Adding an app-nav(Switcher) button on the upper right on the global nav bar. 
Issues to handled:

1. The app-nav button is visible in the top right of the screen in the global header.
2. Clicking the app-nav button opens a drop-down as shown in the design (The options available in the drop-down should be driven by extension slots - how we register things in that slot is handled by a different ticket, [MF-465](https://issues.openmrs.org/browse/MF-465)).
3. Clicking "X" should close the menu. 

Here is the link to the issue on Jira [MF-430](https://issues.openmrs.org/browse/MF-430).

### CODE REVIEW

File: src/components/navbar/navbar.component.tsx [hunk 1]
```diff
@@ -55,6 +57,18 @@ const Navbar: React.FC<NavbarProps> = ({
     }
   };
 
+  const [switcherToggler, setSwitcherToggler] = React.useState(false);
+
+  const handleSwitcherClick = React.useCallback(() => {
+    setSwitcherToggler(switcherToggler => !switcherToggler);
+  }, []);
+
+  const icon = switcherToggler ? (
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: Could be shorter if you just follow my previous comment.

```
const Icon = switchToggler ? Close20 : AppSwitcher20;

// later
<Icon onClick={handleSwitcherClick} />
```

File: src/components/navbar/navbar.component.tsx [hunk 2]
```diff
@@ -94,6 +108,25 @@ const Navbar: React.FC<NavbarProps> = ({
                 >
                   <UserAvatarFilledAlt20 />
                 </HeaderGlobalAction>
+                <HeaderGlobalAction
+                  aria-label="app switcher"
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: These things are for screen readers and they should be filled out carefully.

`aria-label` -> `Toggle App Menu`

`aria-labelledby` -> ??? where is "app swithcer Icon" (note your typo) defined? This is referring to some ID. If there is none then don't use it.

File: src/components/navbar/navbar.component.tsx [hunk 3]
```diff
@@ -94,6 +108,25 @@ const Navbar: React.FC<NavbarProps> = ({
...
+                  aria-labelledby="app swithcer Icon"
+                  name="switcher"
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: Why do we have (and need) a `name` here?

File: src/components/navbar/navbar.component.tsx [hunk 4]
```diff
@@ -94,6 +108,25 @@ const Navbar: React.FC<NavbarProps> = ({
...
+                >
+                  {icon}
+                </HeaderGlobalAction>
+                <HeaderPanel
+                  aria-label="Header Panel"
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: This label does not tell anything.

File: src/components/navbar/navbar.component.tsx [hunk 5]
```diff
@@ -94,6 +108,25 @@ const Navbar: React.FC<NavbarProps> = ({
...
+                  style={{
+                    display: switcherToggler ? "block" : "none",
+                    height: "12rem",
+                    overFlow: "overflow"
+                  }}
+                  expanded
+                >
+                  <ExtensionSlot extensionSlotName="global-nav-menu-slot"></ExtensionSlot>
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: What's the difference between `nav-menu-slot` (which we have already) and `global-nav-menu-slot`. The former is also globally available.

Can you give a better name? What's the purpose of this panel? What kind of items should be in there?

Also please self-close empty elements for readability.

File: src/components/navbar/navbar.component.tsx (hunk 1 above)
Comment (joachim-junior) [reply src/components/navbar/navbar.component.tsx]: Sir, I found difficulties implementing it like this. I got some typescript errors. I tried figuring out how to do this. But couldn't implement it exactly like this.

File: src/components/navbar/navbar.component.tsx (hunk 5 above)
Comment (joachim-junior) [reply src/components/navbar/navbar.component.tsx]: Sir, the Items in it I'm not too sure of. So I asked Ms Grace for clarifications. Still waiting on her. But the designed showed some options already in implemented in the nav-meu-slot.

File: src/components/navbar/navbar.component.tsx (hunk 1 above)
Comment (FlorianRappl) [reply src/components/navbar/navbar.component.tsx]: It works. Just tell TS that this is a `React.FC<any>` or specify the type of the props (but I guess its anyway `any` from the beginning on).

Also make sure to have it named like I have. Know your JSX rule (e.g., `Icon` not `icon`).

Comment (joachim-junior) [reply src/components/navbar/navbar.component.tsx]: I have Implemented it Sir. Thank you very much for guiding me along. I have learnt Alot.

File: src/components/navbar/navbar.component.tsx [hunk 6]
```diff
@@ -55,6 +57,20 @@ const Navbar: React.FC<NavbarProps> = ({
     }
   };
 
+  const [switcherToggler, setSwitcherToggler] = React.useState(false);
+
+  const Icon: React.FC<any> = () => {
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: No. You should *only* have `Icon = switcherToggler ? Close20 : AppSwitcher20`.

Right now you define a new component (and you do that on every render). Also what's up with that style? Why is there a background color hardcoded in there?

Comment (FlorianRappl) [reply src/components/navbar/navbar.component.tsx]: (Additionally that code is wrong, too, as no props are forwarded...).

**Review by FlorianRappl — CHANGES_REQUESTED** [review]

File: src/components/navbar/navbar.component.tsx [hunk 7]
```diff
@@ -94,6 +104,24 @@ const Navbar: React.FC<NavbarProps> = ({
...
+                <HeaderPanel
+                  aria-label="Toggle App Menu Panel"
+                  style={{
+                    display: switcherToggler ? "block" : "none",
+                    height: "12rem",
+                    overFlow: "overflow"
+                  }}
+                  expanded
+                >
+                  <ExtensionSlot extensionSlotName="toggle-app-menu-slot"></ExtensionSlot>
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: Just `app-menu-slot` (the toggle part is really not interesting, nor may it be accurate as it reflects the current state of the implementation / behavior).

Also this is an empty element; just self-close it.

**Review by FlorianRappl — COMMENTED** [review]
I think we are close to target. Thanks!

**Comment by joachim-junior** [comment]
Hi @FlorianRappl, please can you have a look now, so that you tell me if there are some changes to be made. Also I don't seem to understand why some tests failed. It should because of some prettier Configurations.

File: src/components/navbar/navbar.component.tsx [hunk 8]
```diff
@@ -93,6 +106,24 @@ const Navbar: React.FC<NavbarProps> = ({
...
+                  <Icon />
+                </HeaderGlobalAction>
+                <HeaderPanel
+                  aria-label="Toggle App Menu Panel"
+                  style={{
+                    display: switcherToggler ? "block" : "none",
+                    height: "12rem",
+                    overFlow: "overflow"
+                  }}
+                  expanded
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: Just a question: Wouldn't toggling `expanded` be better than toggle between block and none?

Comment (joachim-junior) [reply src/components/navbar/navbar.component.tsx:122]: Hi @FlorianRappl , I tried this ->  **_expanded = switcherToggler ? 'expanded' : ''_** . But it didn't work. I don't if this is what you want me to do.

Comment (joachim-junior) [reply src/components/navbar/navbar.component.tsx]: Hi @FlorianRappl, Is everything okay now. I'm actually excited about this learning process. Is there any other I need to change ?

File: src/components/navbar/navbar.component.tsx [hunk 9]
```diff
@@ -16,8 +16,7 @@ import {
   HeaderPanel
 } from "carbon-components-react/es/components/UIShell";
 import { LoggedInUser, UserSession } from "../../types";
-import styles from "./navbar.scss";
-
+import { Switcher } from "carbon-components-react";
```
Comment (FlorianRappl) [inline src/components/navbar/navbar.component.tsx]: Import from the respective submodule.

**Review by FlorianRappl — APPROVED** [review]
LGTM!

**Pull request merged**